### PR TITLE
KFSPTS-10464 prevent CBCP route errors from stopping next cbcp docs from routing, and adding report

### PR DESCRIPTION
**Pull request by jhulslander**
Please note there is a related PR in the non prod sql repo

### CODE REVIEW

File: src/main/java/edu/cornell/kfs/fp/batch/service/impl/CorporateBilledCorporatePaidCreateDocumentServiceImpl.java [hunk 1]
```diff
@@ -10,14 +10,14 @@
...
 import org.kuali.kfs.fp.businessobject.ProcurementCardDefault;
-import org.kuali.kfs.fp.businessobject.ProcurementCardSourceAccountingLine;
-import org.kuali.kfs.fp.businessobject.ProcurementCardTargetAccountingLine;
 import org.kuali.kfs.fp.businessobject.ProcurementCardTransaction;
 import org.kuali.kfs.fp.businessobject.ProcurementCardTransactionDetail;
 import org.kuali.kfs.fp.document.ProcurementCardDocument;
 import org.kuali.kfs.krad.bo.AdHocRouteRecipient;
 import org.kuali.kfs.krad.exception.ValidationException;
 import org.kuali.kfs.krad.service.DataDictionaryService;
+import org.kuali.kfs.krad.util.ErrorMessage;
```
Comment (ajd299) [inline src/main/java/edu/cornell/kfs/fp/batch/service/impl/CorporateBilledCorporatePaidCreateDocumentServiceImpl.java]: This import seems to no longer be used.

File: src/main/java/edu/cornell/kfs/fp/batch/service/impl/CorporateBilledCorporatePaidRouteStepReportServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+        }
+    }
+
+    protected String buildValidationErrorMessageForSingleError(ErrorMessage errorMessage) {
+        String errorMessageString = configurationService.getPropertyValueAsString(errorMessage.getErrorKey());
+        if (StringUtils.isBlank(errorMessageString)) {
+            throw new RuntimeException("Cannot find error message for key: " + errorMessage.getErrorKey());
+        }
+        
+        Object[] messageParameters = (Object[]) errorMessage.getMessageParameters();
```
Comment (ajd299) [inline src/main/java/edu/cornell/kfs/fp/batch/service/impl/CorporateBilledCorporatePaidRouteStepReportServiceImpl.java]: Is the cast to Object[] from String[] needed? 
Also, is using "!ObjectUtils.IsNull(messageParameters)" preferred over "!= null" for an array of objects?

File: src/main/java/edu/cornell/kfs/fp/batch/service/impl/CorporateBilledCorporatePaidRouteStepReportServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,136 @@
+package edu.cornell.kfs.fp.batch.service.impl;
+
+import java.io.File;
+import java.text.MessageFormat;
+import java.util.ArrayList;
```
Comment (ajd299) [inline src/main/java/edu/cornell/kfs/fp/batch/service/impl/CorporateBilledCorporatePaidRouteStepReportServiceImpl.java]: There are 2 unused imports here (File and ArrayList)

File: src/main/java/edu/cornell/kfs/fp/CuFPKeyConstants.java [hunk 4]
```diff
@@ -5,4 +5,9 @@
     public static final String ERROR_CBCP_BATCH_UPLOAD_EMAIL_SUBJECT = "error.cbcp.batch.upload.email.subject";
     public static final String ERROR_CBCP_BATCH_UPLOAD_EMAIL_BODY_STARTER = "error.cbcp.batch.upload.email.body.starter";
     public static final String ERROR_CREATE_ACCOUNTING_DOCUMENT_ATTACHMENT_DOWNLOAD = "error.create.accounting.document.attachment.download";
+    public static final String MESSAGE_CBCP_ROUTE_STEP_REPORT_SUPMMARY_SUBHEADER = "message.cbcp.route.step.report.summary.subheader";
+    public static final String MESSAGE_CBCP_ROUTE_STEP_REPORT_SUPMMARY_DETAIL_LINE = "message.cbcp.route.step.report.summary.detail.line";
+    public static final String MESSAGE_CBCP_ROUTE_STEP_REPORT_ERROR_SUBHEADER = "message.cbcp.route.step.report.error.subheader";
+    public static final String MESSAGE_CBCP_ROUTE_STEP_REPORT_ERROR_DETAIL_LINE = "message.cbcp.route.step.report.error.detail.line";
+    public static final String MESSAGE_CBCP_ROUTE_STEP_REPORT_NO_ERRORS = "message.cbcp.route,step.report.no.errors";
```
[the remark this answers is not in the document]
Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/fp/CuFPKeyConstants.java]: ah, good catch, I shall fix

File: src/main/java/edu/cornell/kfs/fp/batch/service/impl/CorporateBilledCorporatePaidCreateDocumentServiceImpl.java (hunk 1 above)
Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/fp/batch/service/impl/CorporateBilledCorporatePaidCreateDocumentServiceImpl.java]: ok

File: src/main/java/edu/cornell/kfs/fp/batch/service/impl/CorporateBilledCorporatePaidRouteStepReportServiceImpl.java (hunk 3 above)
Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/fp/batch/service/impl/CorporateBilledCorporatePaidRouteStepReportServiceImpl.java]: ah, missed them, shall remove

File: src/main/java/edu/cornell/kfs/fp/batch/service/impl/CorporateBilledCorporatePaidRouteStepReportServiceImpl.java (hunk 2 above)
Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/fp/batch/service/impl/CorporateBilledCorporatePaidRouteStepReportServiceImpl.java]: I reused this code from another part of KFS.  I think using CollectionUtils.isNotEmpty would be best here.  I shall switch it.  Thanks!

Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/fp/batch/service/impl/CorporateBilledCorporatePaidRouteStepReportServiceImpl.java]: I reused this code from another part of KFS.  I think using CollectionUtils.isNotEmpty would be best here.  I shall switch it.  Thanks, I shall clean  it up

**Comment by jhulslander** [comment]
retest this please

**Comment by jhulslander** [comment]
this is ready for review again. it looks like the unit tests aren't running yet though.  If changes are needed, I can tackle them Thursday

**Comment by ajd299** [comment]
retest this please

**Pull request merged**